From a3b863775176d764a6dc30720bded9ff80c5b12a Mon Sep 17 00:00:00 2001
From: Mohit Sabharwal <mohit@cloudera.com>
Date: Mon, 11 Jul 2016 18:32:52 -0700
Subject: [PATCH 0669/1164] CDH-40809 : HIVE-13862 (Reverts CLOUDERA-BUILD:
 CDH-42090) : Fix exception in direct SQL check
 for getNumPartitionsViaSqlFilter

This commit also reverts CLOUDERA-BUILD: CDH-42090: getNumPartitionsByFilter fail on direct-sql due to datanucleus old version"

Change-Id: Ibe261eb0c694702c5483301fa83f1aab89dcfa5f
---
 .../hadoop/hive/metastore/MetaStoreDirectSql.java  |   12 ++++--------
 1 file changed, 4 insertions(+), 8 deletions(-)

diff --git a/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java b/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java
index 88fbb7d..0e8a595 100644
--- a/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java
+++ b/metastore/src/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java
@@ -838,18 +838,14 @@ public int getNumPartitionsViaSqlFilter(SqlFilterForPushdown filter) throws Meta
 
     long start = doTrace ? System.nanoTime() : 0;
     Query query = pm.newQuery("javax.jdo.query.SQL", queryText);
-
-    List<Integer> result = executeWithArray(query, params, queryText);
-    if (result == null || result.size() == 0) {
-      throw new MetaException("Could not get number of partitions from Direct SQL.");
-    }
-
+    query.setUnique(true);
+    @SuppressWarnings("unchecked")
+    int sqlResult = extractSqlInt(query.executeWithArray(params));
     long queryTime = doTrace ? System.nanoTime() : 0;
     timingTrace(doTrace, queryText, start, queryTime);
-    return extractSqlInt(result.get(0));
+    return sqlResult;
   }
 
-
   private void timingTrace(boolean doTrace, String queryText, long start, long queryTime) {
     if (!doTrace) return;
     LOG.debug("Direct SQL query in " + (queryTime - start) / 1000000.0 + "ms + " +
-- 
1.7.9.5

